Theory-on-regression.md

```diff
@@ -0,0 +1,97 @@
+$$\hat{y}(w,x) = w_0 + w_1x_1$$
+it's also called univariate linear regression.
+
+### Loss function
+
+Loss function is to measure the "error" between real values and predicted values. For regression, it can be presented as the following form:
+
+$$ J(w) = ||y - \hat{y}||^2 $$
+
+which is the sum of squares of the differences between real values $y$ and predicted values $\hat{y}$.
+
+### Optimization algorithm
+
+Optimization algorithm is the algorithms used to minimize the loss function, i.e. to keep changing weights $w$ to reduce loss $J(w)$ until it hopefully ends up at a minimum.
+
+The most basic and popular optimization algorithm is [gradient descent](https://en.wikipedia.org/wiki/Gradient_descent), also known as steepest descent. It can be described as below:
+
+$$ \color{gray} \text{repeat until convergence: } \\
+     \color{black} w_i := w_i - \alpha\frac{\partial{}}{\partial{w}}J(w) \\
+    \color{gray}
+    \text{for } i = (0, \cdots, n) \\
+    \alpha \text{ is learning rate}
+$$
+
+The basic idea is to take repeated steps in the opposite direction of the gradient of the loss function, i.e. $\displaystyle -\frac{\partial{}}{\partial{w}}J(w)$, which will lead to a local minimum of loss function, as shown below.
+
+![Gradient descent 1D](image/Gradient-descent.png)
+<!-- Figure Gradient-descent.png from https://imaddabbura.github.io/img/gradient-descent-algorithms/gradients.PNG -->
+
+The process of gradient descent for two features would look like the dynamic diagram below:
+![Gradient descent 2D dynamic](image/Gradient_descent_dynamic.gif)
+<!-- Figure Gradient_descent_dynamic.gif from https://blog.paperspace.com/intro-to-optimization-in-deep-learning-gradient-descent/ -->
+
+
+## Neural network (non-linear regression)
+
+### Model representation
+
+![Neural network](image/Neural_network.png)
+
+The diagram demonstrates one hidden layer neural network.
+
+<!-- Text from https://scikit-learn.org/stable/modules/neural_networks_supervised.html#multi-layer-perceptron -->
+The input layer consists of a set of neurons representing the input features.
+
+ Each neuron in the hidden layer transforms the values from the previous layer with a weighted linear summation $w_1x_1 + w_2x_2 + \cdots + w_nx_n + bias$, followed by a non-linear activation function, e.g. the [logistic or sigmoid function](https://en.wikipedia.org/wiki/Sigmoid_function).
+
+ The output layer receives the values from the last hidden layer and transforms them into output values with linear function.
+
+The neural network available in `scikit-learn` is just look like the architecture in the diagram, i.e. multi-layer perceptron. To use more complex neural network models, deep learning framework should be used, e.g. [PyTorch, Keras, TensorFlow, etc](https://en.wikipedia.org/wiki/Comparison_of_deep_learning_software). Want to have a look at various neural network models? Try https://alexlenail.me/NN-SVG/index.html.
+
+#### Activation function
```

Comment:
I don't think it's very relevant to show the different activation functions, we can simply note that we need a non-linear function there and what is the most common choice.

Reply:
It's interesting to add here what the benefits and drawbacks are of neural networks over linear regression. E.g.:

`The design of the neural network, with it's non-linearity and layers stacked on top of each other, allows for much more complex patterns to be detected than with linear regression. We can view the neurons as 'feature detectors', that retrieve relevant pieces of information from the input data. The flipside of the coin, is that often we need large amounts of training data to be able to learn these features.`

Reply:
> I don't think it's very relevant to show the different activation functions, we can simply note that we need a non-linear function there and what is the most common choice.

Not fully agree. `scikit-learn` has the parameter to set activation function, it's nice to briefly  mention these functions here, of course we don't want to go details. And you're right, the most important is let them know the most common choice.

Rephrased text a bit.